File: executionmodel/src/FilterLibrary.cpp
#include "FilterLibrary.h"
#include <iostream>

using namespace std;

FilterLibrary::FilterLibrary(std::string libName) : library(libName), libName(libName) {
	// Open the library
	if(!library.open()) {
		cerr << "Failed to open library " << libName << endl;
		valid = false;
		return;
	}

	// Get all the function pointers
	getFilterLibraryVendor = library.get<char*>("getFilterLibraryVendor");
	getFilterLibraryName = library.get<char*>("getFilterLibraryName");
	getFilterLibraryVersion = library.get<char*>("getFilterLibraryVersion");
	getRegisteredFilterClassCount = library.get<int>("getRegisteredFilterClassCount");
	getRegisteredFilterClassName = library.get<char*, int>("getRegisteredFilterClassName");
	buildFilterFunc = library.get<Filter*, const char*, const char*>("buildFilter");

	if(!getFilterLibraryVendor || !getFilterLibraryName || !getFilterLibraryVersion || !getRegisteredFilterClassCount || !getRegisteredFilterClassName || !buildFilterFunc) {
		cerr << "Library " << libName << " does not export all required functions" << endl;
		valid = false;
		return;
	}

	vendor = getFilterLibraryVendor();
	name = getFilterLibraryName();
	version = getFilterLibraryVersion();

	// Load all the filter class names
	const int registeredFilterClassCount = getRegisteredFilterClassCount();
	for(int i = 0; i < registeredFilterClassCount; ++i) {
		string filterClassName(getRegisteredFilterClassName(i));
		filterClassNames.push_back(filterClassName);
	}

	valid = true;
}

FilterLibrary::~FilterLibrary() {}

File: javanativebridge/src/main/java/im/jeanfrancois/openvp/nativebridge/FilterLibrary.java
package im.jeanfrancois.openvp.nativebridge;

import java.util.ArrayList;
import java.util.List;

/**
 * Object representation of a filter library.
 *
 * @author jfim
 */
public class FilterLibrary {
	private String vendor;
	private String name;
	private String version;
	private String libraryName;
	private int libraryIndex;
	private List<String> filterClasses = new ArrayList<String>();

	FilterLibrary(int libraryIndex, ExecutionModelLibrary executionModelLibrary) {
		this.libraryIndex = libraryIndex;
		vendor = executionModelLibrary.getFilterLibraryVendor(libraryIndex);
		name = executionModelLibrary.getFilterLibraryName(libraryIndex);
		libraryName = executionModelLibrary.getFilterLibraryLibraryName(libraryIndex);
		version = executionModelLibrary.getFilterLibraryVersion(libraryIndex);

		// Load filter classes
		final int filterClassCount = executionModelLibrary.getRegisteredFilterClassCount(libraryIndex);
		for(int i = 0; i < filterClassCount; ++i) {
			filterClasses.add(executionModelLibrary.getRegisteredFilterClassName(libraryIndex, i));
		}
	}

	public String getVendor() {
		return vendor;
	}

	public String getName() {
		return name;
	}

	public String getVersion() {
		return version;
	}

	public String getLibraryName() {
		return libraryName;
	}

	public List<String> getFilterClasses() {
		return filterClasses;
	}
}
